File: python/Four-Square.py
# -*- coding: utf-8 -*-
#
# Four-Square Cipher
# 
# @author  lellansin <lellansin@gmail.com>
# @website http://www.lellansin.com/tutorials/ciphers
#
import re

# 默认密表
table = [['A', 'B', 'C', 'D', 'E'], 
         ['F', 'G', 'H', 'I', 'J'], 
         ['K', 'L', 'M', 'N', 'O'], 
         ['P', 'R', 'S', 'T', 'U'], 
         ['V', 'W', 'X', 'Y', 'Z']] 

# 
# 生成棋盘
# 
def generate_table(key = ''):
    # wiki原文：usually omitting "Q" or putting both "I" and "J" in the same location to reduce the alphabet to fit
    alphabet = 'ABCDEFGHIJKLMNOPRSTUVWXYZ'
    table = [[0] * 5 for row in range(5)]
    key = re.sub(r'[\W]', '', key).upper()

    for row in range(5):
        for col in range(5):
            if len(key):
                table[row][col] = key[0]
                alphabet = alphabet.replace(key[0], '')
                key = key.replace(key[0], '')
            else:
                table[row][col] = alphabet[0]
                alphabet = alphabet[1:]
    return table

# 
# 加密
# 
def encrypt(keys, words):
    ciphertext = ''
    words = re.sub(r'[\W]', '', words).upper().replace('Q', '')
    R, L  = generate_table(key[0]), generate_table(key[1])

    for i in range(0, len(words), 2):
        digraphs = words[i:i+2]
        ciphertext += mangle(R, L, digraphs)
    return ciphertext


def mangle(R, L, digraphs):
    a = position(table, digraphs[0])
    b = position(table, digraphs[1])
    return R[a[0]][b[1]] + L[b[0]][a[1]]

# 
# 解密
# 
def decrypt(keys, words):
    ciphertext = ''
    words = re.sub(r'[\W]', '', words).upper().replace('Q', '')
    R, L = generate_table(key[0]), generate_table(key[1])

    for i in range(0, len(words), 2):
        digraphs = words[i:i+2]
        ciphertext += unmangle(R, L, digraphs)
    return ciphertext.lower()

def unmangle(R, L, digraphs):
    a = position(R, digraphs[0])
    b = position(L, digraphs[1])
    return table[a[0]][b[1]] + table[b[0]][a[1]]

# todo
def position(table, ch):
    for row in range(5):
        for col in range(5):
            if table[row][col] == ch:
                return (row, col)
    return (None, None)


if __name__ == '__main__':
    # 本例推算见 http://en.wikipedia.org/wiki/Four-square_cipher

    # 明文
    plaintext = 'help me obiwankenobi'

    # 密匙
    key = ['example', 'keyword']

    # 加密
    ciphertext = encrypt(key, plaintext)
    print(ciphertext)

    # 解密
    print(decrypt(key, ciphertext))

File: javascript/Four-Square.js
/*
 * Four-Square Cipher
 *
 * @author  lellansin <lellansin@gmail.com>
 * @website http://www.lellansin.com/tutorials/ciphers
 */
(function() {

    var FourSquare = {};

    // 默认密表
    var table = 'ABCDEFGHIJKLMNOPRSTUVWXYZ';

    /*
     * 生成棋盘
     */
    FourSquare.generateTable = function(key) {
        var alphabet = 'ABCDEFGHIJKLMNOPRSTUVWXYZ';
        var table = '';
        key = key || '';
        key = key.replace(/[\W]/, '').toUpperCase();

        for (var row = 0; row < 5; row++) {
            for (var col = 0; col < 5; col++) {
                if (key.length) {
                    table += key[0]
                    alphabet = alphabet.replace(key[0], '')
                    key = key.replace(new RegExp(key[0], 'g'), '')
                } else {
                    table += alphabet[0];
                    alphabet = alphabet.substring(1);
                }
            }
        }
        return table;
    };

    /*
     * 打印棋盘
     */
    FourSquare.printTable = function(table) {
        var str = '  1 2 3 4 5 \n';
        for (var y = 0; y < 5; y++) {
            str += (y + 1) + ' ';
            for (var x = 0; x < 5; x++) {
                str = str + table[y * 5 + x] + ' ';
            }
            str += '\n';
        }
        console.log(str);
    };

    /*
     * 加密
     */
    FourSquare.encrypt = function(keys, text) {
        var R = FourSquare.generateTable(key[0]);
        var L = FourSquare.generateTable(key[1]);

        return text
            .toUpperCase()
            .replace(/[\WQ]/g, '')
            .replace(/[\w]{1,2}/g, function(digraphs) {
                var a = position(table, digraphs[0]);
                var b = position(table, digraphs[1] || 'X');
                return R[a[0] * 5 + b[1]] + L[b[0] * 5 + a[1]];
            });
    };

    /*
     * 解密
     */
    FourSquare.decrypt = function(keys, text) {
        var R = FourSquare.generateTable(key[0]);
        var L = FourSquare.generateTable(key[1]);

        return text
            .toUpperCase()
            .replace(/[\WQ]/g, '')
            .replace(/[\w]{1,2}/g, function(digraphs) {
                var a = position(R, digraphs[0]);
                var b = position(L, digraphs[1] || 'X');
                return table[a[0] * 5 + b[1]] + table[b[0] * 5 + a[1]];
            });
    };

    var position = function(table, ch) {
        var index = table.indexOf(ch);
        return [Math.floor(index / 5), index % 5];
    }

    /* -------------------- 测试 --------------------- */
    // 本例推算见 http://en.wikipedia.org/wiki/Four-square_cipher

    // 明文
    var text = 'help me obiwankenobi';

    // 密匙
    var key = ['example', 'keyword'];

    // 加密
    var ciphertext = FourSquare.encrypt(table, text);
    console.log(ciphertext);

    // 解密
    console.log(FourSquare.decrypt(table, ciphertext));

}());
